util/clustering/kmeans_clustering.py

```diff
@@ -0,0 +1,231 @@
+    feat = np.array(list(data.values()))
+
+    # reshape so that there are 210 samples of 4096 vectors
+    feat = feat.reshape(-1,4096)
+
+
+    # # reduce the amount of dimensions in the feature vector
+    print('------------ feature reduction ------------')
+    pca = PCA(n_components=100, random_state=22)
+    x = pca.fit_transform(feat)
+
+    os.chdir(file_dir)
+
+    # this is just incase you want to see which value for k might be the best 
+    # note: this didn't really work very well (gave best number of clusters = 5
+    # which is obviously too little)
+    # sse = []
+    # list_k = list(range(5, 1000))
+    # max_diff = 0
+    # prev_sse = 0
+    # best_k = 0
+    # attempts = 0
+
+    # print('------------ checking clusters ------------')
+
+    # find the number of clusters that produces the greatest change in sse. 
+    # for k in tqdm(list_k):
+    #     km = KMeans(n_clusters=k, random_state=22)
+    #     km.fit(x)
+    #     inertia = km.inertia_
+    #     if abs(inertia - prev_sse) > max_diff:
+    #         print('new best k : ' + str(k))
+    #         max_diff = inertia 
+    #         best_k = k 
+    #         attempts = 0
+    #     prev_see = inertia
+    #     sse.append(inertia)
+    #     attempts += 1
+
+    # print('best number of clusters: ' + str(best_k))
+
+    # # Plot sse against k
+    # print('------------ plotting sse against k ------------')
+    # plt.figure(figsize=(6, 6))
+    # plt.plot(list_k, sse)
+    # plt.xlabel(r'Number of clusters *k*')
+    # plt.ylabel('Sum of squared distance')
+    # plt.show()
+    # plt.savefig(args.output_data_dir + '/cluster_analysis.png')
+
+    # num_clusters = best_k
+
+    # cluster feature vectors
+    kmeans = KMeans(n_clusters=150, random_state=22)
+    label = kmeans.fit_predict(x)
+    kmeans.fit(x)
+    
+    print('------------ printing labels ------------')
+    print(kmeans.labels_)
+    #Getting unique labels
+    
+    u_labels = np.unique(kmeans.labels_)
+    
+    #plotting the results:
+    
+    for i in u_labels:
+        plt.scatter(x[label == i , 0] , x[label == i , 1] , label = i)
+    plt.savefig(args.output_data_dir + '/pca_clustering.png', bbox_inches='tight')
+    plt.close()
+
+    # holds the cluster id and the images { id: [images] }
+    print('------------ getting groups ------------')
+
+    groups = {}
+    for file, cluster in zip(filenames,kmeans.labels_):
+        if cluster not in groups.keys():
+            groups[cluster] = []
+            groups[cluster].append(file)
+        else:
+            groups[cluster].append(file)
+    print(groups)
+
+    print('------------ creating label files ------------')
+
+
+    for cluster in groups.keys(): 
+        with open(f"labels/label_{cluster}.txt", "w") as output:
+            for file in groups[cluster]:
+                output.write('%s\n' % file)
+
+    # function that lets you view a cluster (based on identifier)        
+    def view_cluster(cluster):
```

Comment:
Again, there isn't a reason for this to be an inner-function. Should move this outside of main()

Reply:
Also, like above, don't chdir